feat(cli): gc can ignore user-verified unauthorized non-cdk stacks

dgandhi62 opened the pull request:
Fixes #640 

The gc command currently fails when encountering AccessDenied errors on non-CDK stacks during garbage collection. This PR adds a new option `--unauth-native-cfn-stacks-to-skip` to handle this situation. 

Users can specify multiple stack names or glob patterns to skip when unauthorized access errors occur. When stacks matching the provided patterns encounter access denied errors, the command shows a confirmation prompt asking whether to skip these stacks and continue with garbage collection. The default response is 'no' to ensure users explicitly confirm the action. It remains the user's responsibility to ensure that the stacks entered are non-cdk stacks. 

The 'no' default was chosen over 'yes' to prevent unintended behavior until a future `--yes` flag is implemented for the cdk-cli. While a shorter alias was considered for the option name, it was rejected to avoid user confusion.


---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license

kaizencc commented on packages/@aws-cdk/toolkit-lib/lib/api/garbage-collection/garbage-collector.ts:
```diff
@@ -178,6 +178,18 @@ interface GarbageCollectorProps {
    */
   readonly confirm?: boolean;
+
+  /**
```
missing documentation for both of these

dgandhi62 replied:
Have updated

kaizencc commented on packages/@aws-cdk/toolkit-lib/lib/api/garbage-collection/stack-refresh.ts:
```diff
@@ -97,9 +120,16 @@ function bootstrapFilter(parameters?: ParameterDeclaration[], qualifier?: string
 }
 
-export async function refreshStacks(cfn: ICloudFormationClient, ioHelper: IoHelper, activeAssets: ActiveAssetCache, qualifier?: string) {
+export async function refreshStacks(
```
i think its time for refreshStacks to take in a property bag

kaizencc commented on packages/aws-cdk/lib/cli/cdk-toolkit.ts:
```diff
@@ -1918,6 +1918,18 @@ export interface GarbageCollectionOptions {
    * @default false
    */
   readonly confirm?: boolean;
+
```
both missing documentation

kaizencc commented on packages/@aws-cdk/toolkit-lib/lib/api/garbage-collection/garbage-collector.ts:
```diff
@@ -178,6 +178,18 @@ interface GarbageCollectorProps {
+   *
+   * @default []
+   */
+  readonly ignoreStacks?: string[];
```
```suggestion
  readonly ignoreNonCdkStackNames?: string[];
```

kaizencc replied:
we need to make it clear that garbage collection is unsafe if we skip CFN stacks created by CDK. to use this property safely, the customer must know beforehand that the stack in question was not created via CDK.

dgandhi62 replied:
Have made the change

kaizencc commented on packages/@aws-cdk/toolkit-lib/lib/api/garbage-collection/garbage-collector.ts:
```diff
@@ -178,6 +178,18 @@ interface GarbageCollectorProps {
+   *
+   * @default false
+   */
+  readonly skipUnauthorizedStacks?: boolean;
```
in fact, i am not comfortable with this property at all. we cannot supply a way for a customer to unknowningly purge assets from an unauthorized CDK stack. since this is a tricky and uncommon situation, i would like us to make sure the user interface is explicit about which stacks we can safely ignore.

dgandhi62 replied:
I've added warnings to the user and mentioned which stacks are being ignored. It isn't garbage collecting on the unauth stacks, instead, just skipping them and not throwing an error

kaizencc commented on packages/aws-cdk/lib/cli/cli.ts:
```diff
@@ -487,6 +487,8 @@ export async function exec(args: string[], synthesizer?: Synthesizer): Promise<n
           createdBufferDays: args['created-buffer-days'],
           bootstrapStackName: args.toolkitStackName ?? args.bootstrapStackName,
           confirm: args.confirm,
+          ignoreStacks: args['ignore-stacks'],
```
missing a cli integ test for this to ensure that this is actually exposed via the cli. integ test should deploy 2 stacks with an asset each, one stack with a prefix, and then verify that gc with ignore stacks actually _does_ delete the asset of the stack with prefix.

kaizencc reviewed (changes requested):
also missing a readme entry. this definitely needs it because it is a very tricky property to add and without the proper documentation, users will use it incorrectly and accidentally delete important assets.

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/api/garbage-collection/garbage-collector.ts:
```diff
@@ -178,6 +178,20 @@ interface GarbageCollectorProps {
+   *
+   * @default []
+   */
+  readonly ignoreNonCdkStacks?: string[];
```
The way this is implemented we are skipping _any_ stack, not just CDK stacks. Like @kaizencc I'm not convinced this is a good idea in the first place, but at leat the option need to be named correctly.

mrgrain reviewed (dismissed):
option names

gandhya committed:
feat(cli): add skip-unauthorized-stacks-when-noncdk option for garbage collection

- Removed old functionality on ignore-stacks & skip-unauthorized-stacks
- Add --skip-unauthorized-stacks-when-noncdk CLI option to handle AccessDenied errors for non-cdk stack names entered by user
- Update interfaces and constructors with new parameter

kaizencc commented on packages/@aws-cdk/toolkit-lib/lib/api/garbage-collection/garbage-collector.ts:
```diff
@@ -178,6 +178,14 @@ interface GarbageCollectorProps {
+   *
+   * @default undefined
+   */
+  readonly skipUnauthorizedStacksWhenNonCdk?: string[];
```
we need to workshop this name. how about:

```ts
{
  readonly nonCdkStacksToSkip?: string[];
}
```

or 

```ts
{
  readonly nativeCloudFormationStacksToSkip?: string[];
}
```

It's very important to get this right. @mrgrain do you have ideas?

kaizencc replied:
i think, while the original user had an unauthorized stack they wanted to skip, the real importance of this property is that it must must must not be a cdk stack. my preference is still `nativeCloudFormationStacksToSkip`. Please follow up with @mrgrain to make a determination here.

kaizencc commented on packages/@aws-cdk/toolkit-lib/lib/api/garbage-collection/garbage-collector.ts:
```diff
@@ -178,6 +178,14 @@ interface GarbageCollectorProps {
+
+  /**
+   * Non-CDK stack names or glob patterns to skip when encountering unauthorized access errors during garbage collection.
+   * You must explicitly specify non-CDK stack names - CDK stacks will be rejected to prevent accidental asset deletion.
```
"CDK stacks will be rejected to prevent accidental asset deletion"

this reads like we are checking for the user whether or not stacks supplied here are CDK stacks.

kaizencc commented on packages/@aws-cdk/toolkit-lib/lib/api/garbage-collection/garbage-collector.ts:
```diff
@@ -178,6 +178,14 @@ interface GarbageCollectorProps {
+   * Non-CDK stack names or glob patterns to skip when encountering unauthorized access errors during garbage collection.
+   * You must explicitly specify non-CDK stack names - CDK stacks will be rejected to prevent accidental asset deletion.
+   *
+   * @default undefined
```
```suggestion
   * @default []
```

dgandhi62 replied:
[**skipUnauthorizedStacksWhenNonCdk: arrayFromYargs(args['skip-unauthorized-stacks-when-noncdk'])**](https://github.com/aws/aws-cdk-cli/pull/850/files#diff-be9d125a91b3869bf888755d6d10cd596d19397a54f3a3bce60dd4c8ee2c0406R490)

arrayFromYargs returns undefined for an empty input. Would [] be technically valid documentation as a default?

kaizencc commented on packages/@aws-cdk/toolkit-lib/lib/api/garbage-collection/stack-refresh.ts:
```diff
@@ -56,20 +80,40 @@ async function fetchAllStackTemplates(cfn: ICloudFormationClient, ioHelper: IoHe
+      // Check if this is a CloudFormation access denied error
+      if (error.name === 'AccessDenied') {
+        if (shouldSkipStack(stack, skipUnauthorizedStacksWhenNonCdk)) {
+          await ioHelper.defaults.warn(
+            `Skipping unauthorized stack '${stack}' as specified in --skip-unauthorized-stacks-when-noncdk`,
+          );
```
instead of printing every time we find an unauthorized stack, why don't we store these skipped stacks in memory and then, after we are done parsing stacks and _before we delete_ we prompt the user with something like "skipped these stacks that YOU specified as not CDK stacks. if they are CDK stacks their assets will be deleted!!! are you sure? y/n/t"

y/n/t is yes/no/trust, which we should have an example of elsewhere in garbage collection since it prompts the user elsewhere too.

dgandhi62 replied:
Latest commit adds this functionality. Will make another commit with a README once the option name is finalized

kaizencc reviewed (changes requested):
also definitely needs a readme update :)

dgandhi62 committed:
feat(gc): improve unauthorized stack handling with batch prompts and CI support

- Replace individual warnings with single batch prompt for unauthorized stacks
- Add CDK_GC_AUTO_APPROVE_UNAUTHORIZED env var for CI/CD automation
- Add CI detection to fail fast instead of hanging on user prompts
- Add performance optimization with early return for empty assets

dgandhi62 committed:
Add environment variable to test for ci environments

dgandhi62 committed:
fix(test): add CI detection for garbage collection tests

- Remove global environment variable from test setup and add it locally to test
- Add CI detection in failing tests

kaizencc commented on packages/@aws-cdk/toolkit-lib/lib/api/garbage-collection/stack-refresh.ts:
```diff
@@ -11,6 +12,9 @@ export class ActiveAssetCache {
   }
 
   public contains(asset: string): boolean {
+    // To reduce computation if asset is null or undefined
+    if (!asset) return false;
+
```
this change seems unrelated? its also wrong -- asset cannot be null or undefined as it is defined as a string. what it is currently doing is saying that if `asset == ''`, return `false`. fine, but document that rather than saying asset is null or undefined.

kaizencc commented on packages/@aws-cdk/toolkit-lib/lib/api/garbage-collection/stack-refresh.ts:
```diff
@@ -31,12 +52,78 @@ async function paginateSdkCall(cb: (nextToken?: string) => Promise<string | unde
+    const message = `Found ${unauthorizedStacks.length} unauthorized stack(s): ${stackList}\nDo you want to skip all these stacks?`;
+
+    // Ask user if they want to proceed
+    const response = await ioHelper.requestResponse({
+      time: new Date(),
+      level: 'info',
+      code: 'CDK_TOOLKIT_I9210',
+      message,
+      data: {
+        stacks: unauthorizedStacks,
+        count: unauthorizedStacks.length,
+        responseDescription: '[y]es/[n]o',
+      },
+      defaultResponse: 'y',
+    });
```
this needs to be defined message in private/messages.ts. see other places where we are calling `requestResponse`.

kaizencc commented on packages/@aws-cdk/toolkit-lib/lib/api/garbage-collection/stack-refresh.ts:
```diff
@@ -31,12 +52,78 @@ async function paginateSdkCall(cb: (nextToken?: string) => Promise<string | unde
+    const response = await ioHelper.requestResponse({
+      time: new Date(),
+      level: 'info',
+      code: 'CDK_TOOLKIT_I9210',
```
this is definitely the wrong code... you need to create your own in messages.ts

kaizencc commented on packages/@aws-cdk/toolkit-lib/lib/api/garbage-collection/stack-refresh.ts:
```diff
@@ -31,12 +52,78 @@ async function paginateSdkCall(cb: (nextToken?: string) => Promise<string | unde
+  }
+
+  try {
+    const stackList = unauthorizedStacks.join(', ');
```
```suggestion
    const stackList = unauthorizedStacks.join(',\n');
```

kaizencc commented on packages/@aws-cdk/toolkit-lib/lib/api/garbage-collection/stack-refresh.ts:
```diff
@@ -31,12 +52,78 @@ async function paginateSdkCall(cb: (nextToken?: string) => Promise<string | unde
+    return;
+  }
+
+  // Auto-approve in CI environments
+  if (process.env.CDK_GC_AUTO_APPROVE_UNAUTHORIZED === 'true') {
+    await ioHelper.defaults.info(`Auto-approving ${unauthorizedStacks.length} unauthorized stack(s) in CI mode`);
+    return;
+  }
+
+  // Detect CI environment and fail fast to prevent hanging
+  const isCI = process.env.CI === 'true' ||
+               process.env.GITHUB_ACTIONS === 'true' ||
+               process.env.GITLAB_CI === 'true' ||
+               process.env.JENKINS_URL !== undefined ||
+               process.env.BUILDKITE === 'true';
+
+  if (isCI) {
+    throw new ToolkitError(
+      `Found ${unauthorizedStacks.length} unauthorized stack(s) in CI environment. ` +
+      'Set CDK_GC_AUTO_APPROVE_UNAUTHORIZED=true to auto-approve or configure --skip-unauthorized-stacks-when-noncdk.',
+    );
+  }
```
none of this is necessary. our ioHost handles this sort of thing for us. basically, we send all output to a global IoHost instance (in the cli, its called `CliIoHost`) and that IoHost decides where (or if) we send information. in the case of a ci environment and `CliIoHost`, we just accept the default.

in this case, you've defined the default as `yes`. is that what we want to do? (i think that's ok, but think about it)

dgandhi62 replied:
The issue is that github kept failing the test if I didn't add this check. Is this expected?

kaizencc reviewed (changes requested):
few more comments. this one is tricky!

dgandhi62 committed:
Merge branch 'aws:main' into feat-ignore-skip-stacks

dgandhi62 committed:
feat(gc): change default to 'no' for unauthorized stacks and rename option to --unauth-native-cfn-stacks-to-skip

- Change default response from 'yes' to 'no' when prompting to skip unauthorized stacks
- Rename parameter from skipUnauthorizedStacksWhenNonCdk to unauthNativeCfnStacksToSkip
- Remove wrong ci/cd detection logic

dgandhi62 committed:
fix: add option to cli-config and remove it from parse-command-line-arguments
